features/123-chat-push-notification/api.md

```diff
@@ -0,0 +1,159 @@
+# Push Notification in Chat
+## Feature Overview
+
+This feature enables developer to send customized push notification via cloud function deployment. The approach is to provide hooks in chat conversation including
+
+* Messages
+  * after\_message\_sent
+  * after\_message\_updated
+  * after\_message\_deleted
+* Typing
+  * typing\_started
+* Conversation
+  * after\_conversation\_created
+  * after\_conversation\_deleted
+  * after\_conversation\_updated
+  * after\_user_entered\_conversation
+  * after\_user_leave\_conversation
+
+In addition, the current JS push notification code snippet is too verbose and can be simplified.
+
+### Task
+- Refactor JS SDK Push Notification Function so that it can be called by developer with less code.  
+	- Original Code: 
+	
+	```javascript
+	var skygear = require('skygear');
+	skygear.push.sendToDevice(
+	  ['2aa4af2a-699a-4e43-8d67-7598757fc7ed'], // Device IDs
+	  {
+	    'apns': {
+	        'aps': {
+	            'alert': {
+	                'title': title,
+	                'body': message,
+	            }
+	        },
+	        'from': 'skygear',
+	        'operation': 'notification',
+	    },
+	    'gcm': {
+	         'notification': {
+	              'title': title,
+	              'body': message,
+	          }
+	    },
+		}
+	);
+	```
+
+	- new Code: 
+
+	```javascript
+	var skygear = require('skygear');
+	skygear.push.sendToDevice(
+	  ['2aa4af2a-699a-4e43-8d67-7598757fc7ed'],
+	  {
+         'title': title,
+         'body': message
+	  }
+	);
+	```
+- Create chat hook annotation in chat SDK
+- Enable chat plugin to call pre-defined hook
+
+### Sample Codes
+
+- Sample Javascript Cloud Function
+
+```javascript
+skygearChat.afterMessageSent((message, conversation) => {
+  var userIds = conversation.participants.map(p => p.user.key.recordID);
+  skygear.push.sendToUser(userIds, {'title': 'Title', 'body': 'Hello World'})
+})
+```
+
+- Python Cloud Function
+
+```python
+
+@chat.after_message_sent
+def after_messange_send_handler(message, conversation):
+   user_ids = [p.id.key for p in conversation.participants]
+   notification = {...}
+	skygear.action.push_users(container, user_ids, notification)
+```
+
+### List of APIs
+
+- Javascript
+	- `skygearChat.afterMessageSent((message, conversation) => {})`
+	- `skygearChat.afterMessageUpdated((message, conversation) => {})`
+	- `skygearChat.afterMessageDeleted((message, conversation) => {})`
+	- `skygearChat.typingStarted((conversation, events) => {})`
+	- `skygearChat.afterConversationCreated((conversation) => {})`
+	- `skygearChat.afterConversationUpdated((conversation) => {})`
+	- `skygearChat.afterConversationDeleted((conversation) => {})`
+	- `skygearChat.afterUserAddedToConversation((conversation) => {})`
+	- `skygearChat.afterUserRemovedFromConversation((conversation) => {})`
```

Comment:
for user added / removed, is it better to also have the user record object as the parameter?